src/server/webSocket.js

```diff
@@ -0,0 +1,237 @@
+    }
+
+    const coreEventListenerList = [
+      {
+        eventName: 'error',
+        listener: this.onWSSError.bind(this),
+      },
+      {
+        eventName: 'close',
+        listener: WebSocketServerCore.onWSSClose,
+      },
+    ];
+
+    [...coreEventListenerList, ...externalEventListenerList].forEach(({ eventName, listener }) => {
+      if (!eventName || !listener) {
+        throw ReferenceError(`
+          event listener must contain: eventName and listener params!
+          Received eventName: "${eventName}", listener: "${listener}"
+        `);
+      }
+
+      this.wss.on(eventName, listener);
+    });
+  }
+
+  static onWSSClose(event) {
+    console.log('WebSocket server closed: ', event);
+  }
+
+  onWSSError(error) {
+    console.error('WebSocket server error: ', error);
+
+    if (!this.mailSent) {
+      mailer.sendMail(`<p>Wystąpił błąd na serwerze WebSocket!<br><output>${JSON.stringify(error)}</output></p>`);
+      this.mailSent = true;
+    }
+  }
+
+  static onWSError(event) {
+    console.error('WebSocket error event: ', event);
+
+    if (!event.errno) {
+      throw event;
+    }
+  }
+
+  static parseWSMessage(event, ws) {
+    let parsedEvent = {};
+
+    try {
+      if (!event) {
+        throw Error('ws empty message event');
+      }
+
+      parsedEvent = JSON.parse(event);
+    } catch (exception) {
+      console.error('ws invalid JSON message: ', event, ' threw exception: ', exception);
+      ws.close(HTTP_STATUS_CODES.WS_CLOSE_ERROR_CODE, 'Message is not valid JSON!');
+
+      return null;
+    }
+
+    if (!parsedEvent.pathname) {
+      console.error('ws empty pathname: ', parsedEvent.pathname);
+      ws.close(HTTP_STATUS_CODES.WS_CLOSE_ERROR_CODE, 'Empty pathname!');
+
+      return null;
+    }
+
+    return parsedEvent;
+  }
+}
+
+class WebSocketServer extends WebSocketServerCore {
+  constructor(httpServer) {
+    super(httpServer);
+
+    this.USER_ID_URL = `${config.protocol}://${config.host}:${config.port.q2a}/user-id`;
+    this.SUBPAGE_VISITORS_AMOUNT_MAP = (() => {
+      // create pure object, without prototype
+      const visitorsMap = Object.create(null);
+
+      return {
+        incrementIfExists(subPageName) {
+          if (!(subPageName in visitorsMap)) {
+            visitorsMap[subPageName] = 0;
+          }
+
+          visitorsMap[subPageName]++;
+        },
+        decrementAndEventuallyRemove(subPageName) {
+          if (!(subPageName in visitorsMap) || visitorsMap[subPageName] === 0) {
+            throw Error(`
+              visitorsMap does not have property subPageName or it's value is 0!
+              visitorsMap entries: "${Object.entries(visitorsMap)}", subPageName: "${subPageName}"
+            `);
+          }
+
+          visitorsMap[subPageName]--;
+
+          if (visitorsMap[subPageName] === 0) {
+            delete visitorsMap[subPageName];
+          }
+        },
+      };
+    })();
```

Comment:
W zasadzie to takie coś jest już zaimplementowane i dostępne chociażby z socket.io - https://socket.io/docs/v4/rooms/. Ale to by wymagało też zmian po stronie klienta, więc może na nowym froncie ;)